src/ray/core_worker/transport/direct_actor_transport.cc

```diff
@@ -39,89 +38,68 @@ Status CoreWorkerDirectActorTaskSubmitter::SubmitTask(
-      for (const auto &entry : iter->second) {
-        const auto &task_id = entry.first;
-        const auto num_returns = entry.second;
-        TreatTaskAsFailed(task_id, num_returns, rpc::ErrorType::ACTOR_DIED);
-      }
-      waiting_reply_tasks_.erase(actor_id);
-    }
-
-    // If there are pending requests, treat the pending tasks as failed.
-    auto pending_it = pending_requests_.find(actor_id);
-    if (pending_it != pending_requests_.end()) {
-      for (const auto &request : pending_it->second) {
-        TreatTaskAsFailed(TaskID::FromBinary(request->task_spec().task_id()),
-                          request->task_spec().num_returns(), rpc::ErrorType::ACTOR_DIED);
-      }
-      pending_requests_.erase(pending_it);
-    }
-  }
-}
-
-void CoreWorkerDirectActorTaskSubmitter::ConnectAndSendPendingTasks(
-    const ActorID &actor_id, std::string ip_address, int port) {
+  // Connect.
   std::shared_ptr<rpc::WorkerTaskClient> grpc_client =
       std::make_shared<rpc::WorkerTaskClient>(ip_address, port, client_call_manager_);
-  RAY_CHECK(rpc_clients_.emplace(actor_id, std::move(grpc_client)).second);
+  auto inserted = rpc_clients_.emplace(actor_id, std::move(grpc_client));
+  auto &rpc_client = inserted.first->second;
 
   // Submit all pending requests.
-  auto &client = rpc_clients_[actor_id];
   auto &requests = pending_requests_[actor_id];
   while (!requests.empty()) {
     auto request = std::move(requests.front());
     auto num_returns = request->task_spec().num_returns();
     auto task_id = TaskID::FromBinary(request->task_spec().task_id());
-    DirectActorAssignTask(*client, std::move(request), actor_id, task_id, num_returns);
+    DirectActorAssignTask(*rpc_client, std::move(request), actor_id, task_id,
+                          num_returns);
     requests.pop_front();
   }
 }
 
+void CoreWorkerDirectActorTaskSubmitter::DisconnectActor(const ActorID &actor_id) {
+  std::unique_lock<std::mutex> guard(mutex_);
+
+  // Disconnect.
+  rpc_clients_.erase(actor_id);
+
+  // For tasks that have been sent and are waiting for replies, treat them
+  // as failed when the destination actor is dead or reconstructing.
+  auto iter = waiting_reply_tasks_.find(actor_id);
+  if (iter != waiting_reply_tasks_.end()) {
+    for (const auto &entry : iter->second) {
+      const auto &task_id = entry.first;
+      const auto num_returns = entry.second;
+      TreatTaskAsFailed(task_id, num_returns, rpc::ErrorType::ACTOR_DIED);
+    }
+    waiting_reply_tasks_.erase(actor_id);
+  }
+
+  // If there are pending requests, treat the pending tasks as failed.
+  auto pending_it = pending_requests_.find(actor_id);
+  if (pending_it != pending_requests_.end()) {
+    for (const auto &request : pending_it->second) {
+      TreatTaskAsFailed(TaskID::FromBinary(request->task_spec().task_id()),
+                        request->task_spec().num_returns(), rpc::ErrorType::ACTOR_DIED);
+    }
+    pending_requests_.erase(pending_it);
```

Comment:
I remember in existing logic we will buffer the pending requests when actor is reconstructing, and only treat these tasks as failed when actor is dead (and cannot be reconstructed). But the new logic might be ok too.

Reply:
Hmm yeah I recall now that I had to change this once we set the task counter according to the counter in the handle, instead of in the direct actor transport. We can think about changing it back later, but I don't think it's a big deal. From the user's perspective, failing all pending tasks might actually be better.